packages/frontend/component/src/hooks/use-resize-observer.ts

```diff
@@ -0,0 +1,75 @@
+import type { ResizeObserverEntry } from '@juggle/resize-observer';
+import { useCallback, useEffect, useRef } from 'react';
+
+type UseResizeObserverCallback = (entity: ResizeObserverEntry) => void;
+
+let _resizeObserver: ResizeObserver | null = null;
+const elementsMap = new WeakMap<Element, UseResizeObserverCallback>();
+
+// for debugging
+(window as any)._resizeObserverElementsMap = elementsMap;
+
+/**
+ * @internal get or initialize the ResizeObserver instance
+ */
+const getResizeObserver = () =>
+  _resizeObserver ??
+  (_resizeObserver = new ResizeObserver(entries => {
+    entries.forEach(entry => {
+      const resizeCallback = elementsMap.get(entry.target);
+      resizeCallback?.(entry);
+    });
+  }));
+
+/**
+ * @internal unobserve an element
+ */
+const unobserve = (element: Element | null) => {
+  if (!element) return;
+  const observer = getResizeObserver();
+  observer.unobserve(element);
+  elementsMap.delete(element);
+};
+
+/**
+ * A hook that observes the resize of an element.
+ * @returns A function that can be used to observe an element.
+ *
+ * ```ts
+ * const elRef = useRef<HTMLDivElement>(null);
+ * const observe = useResizeObserver();
+ *
+ * useEffect(() => {
+ *  const dispose = observe(elRef.current, (entry) => {
+ *    console.log(entry.contentRect);
+ *  });
+ *
+ *  return () => dispose();
+ * }, [observe]);
+ * ```
+ */
+export const useResizeObserver = () => {
+  const prevElRef = useRef<Element | null>(null);
+
+  const observe = useCallback(
+    (element: Element, callback: UseResizeObserverCallback) => {
+      unobserve(prevElRef.current);
+
+      const observer = getResizeObserver();
+      prevElRef.current = element;
+      observer.observe(element);
+      elementsMap.set(element, callback);
```

Comment:
Is it possible to have multiple callbacks for the same element?

Reply:
Yes, it is not currently supported. I think it's not a common scenario, or it could be handled externally. What do you think? @pengx17  @EYHN

Reply:
It is technically possible, also `observe` can be called in the same component multiple times on different element and this misusage could hard to be avoided:

```tsx
const Foo = () => {
  const observe = useResizeObserver();
  observe(ref0.current);
  observe(ref1.current);
  return <div0  ref={ref0}/><div1 ref={ref1}/>
}
```

Reply:
I guess the reason of why you brought in `prevElRef` & `elementsMap` is to make sure element are unobserved in the end, but I think it we can make it the user's reponsibility to do cleanup. Then we do not need to worry about the issues I pointed out here.

```
const observe = useCallback(
  (element: Element, callback: UseResizeObserverCallback) => {
    const observer = getResizeObserver();
    observer.observe(element);
    return () => {
      observer.unobserve(element);
    }
  },
  []
);

```

Reply:
Got it, and if so without cleanup, a normal function rather than hook is suitable here.

Reply:
Extracted into a regular method, and now supports binding multiple listeners to the same element. Please take another look when you have time.